### PR TITLE
REGR: Notebook (html) repr of DataFrame no longer follows min_rows/max_rows settings 

### LINKED ISSUE #37359: REGR: Notebook (html) repr of DataFrame no longer follows min_rows/max_rows settings
On master, the html repr in notebooks of a DataFrame is no longer following the min_rows/max_rows logic (see https://pandas.pydata.org/docs/dev/user_guide/options.html#frequently-used-options). 

If you have a DataFrame with many rows, it incorrectly shows 60 (max_rows) rows in the truncated repr intead of 10 (min_rows). 

cc @ivanovmg (might be related to the refactoring you have been doing, but to be clear, I didn't check if this is actually the case!)

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #37359
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Fix display logic in number of rows,
explained here https://pandas.pydata.org/docs/dev/user_guide/options.html#frequently-used-options

### CODE REVIEW

**Comment by ivanovmg** [comment]
@jorisvandenbossche can you please confirm it works? I am still unable to install dev environment (even conda).

**Comment by ivanovmg** [comment]
I managed to install dev env with jupyter. Looks like everything works. But please check in your environment as well.

image

**Comment by jorisvandenbossche** [comment]
I can confirm this works, thanks!

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -633,20 +633,30 @@ def _calc_max_cols_fitted(self) -> Optional[int]:
...
+            if self._is_screen_short(height):
+                max_rows = height
+            else:
+                max_rows = self.max_rows
         else:
             max_rows = self.max_rows
 
+        return self._adjust_max_rows(max_rows)
+
+    def _adjust_max_rows(self, max_rows: Optional[int]) -> Optional[int]:
```
Comment (jorisvandenbossche) [inline pandas/io/formats/format.py:652]: I don't think it is worth creating a separate function for those few lines, I think it fits fine in the function above, as it was before

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -633,20 +633,30 @@ def _calc_max_cols_fitted(self) -> Optional[int]:
...
-        if self._is_screen_short(height):
-            max_rows = height
+        if self._is_in_terminal():
+            _, height = get_terminal_size()
+            if self.max_rows == 0:
+                # rows available to fill with actual data
+                return height - self._get_number_of_auxillary_rows()
+
+            max_rows: Optional[int]
+            if self._is_screen_short(height):
```
Comment (jorisvandenbossche) [inline pandas/io/formats/format.py:643]: What is this screen_short doing? I think it is rather about the dataframe being longer than the screen, and that we want to show all rows possible

File: pandas/tests/io/formats/test_format.py [hunk 3]
```diff
@@ -2029,6 +2029,35 @@ def test_period(self):
...
+        See description here:
+        https://pandas.pydata.org/docs/dev/user_guide/options.html#frequently-used-options
+        """
+        formatter = fmt.DataFrameFormatter(
+            DataFrame(np.random.rand(length, 3)),
+            max_rows=max_rows,
+            min_rows=min_rows,
+        )
+        result = formatter.max_rows_fitted
+        assert result == expected
```
Comment (jorisvandenbossche) [inline pandas/tests/io/formats/test_format.py:2059]: This is actually already tested (see https://github.com/pandas-dev/pandas/issues/37359#issuecomment-715201488, so not sure it is necessarily needed to add additional tests). But so the problem is that it was not catching the regression (and I suppose those tests above will also not have catched it?)

I will try to think about another way to test it

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the quick follow-up!

File: pandas/tests/io/formats/test_format.py (hunk 3 above)
Comment (ivanovmg) [reply pandas/tests/io/formats/test_format.py:2059]: These tests do fail on master

Comment (ivanovmg) [reply pandas/tests/io/formats/test_format.py:2059]:
```
pandas\tests\io\formats\test_format.py:2059: AssertionError
============================================================== short test summary info ===============================================================
FAILED pandas/tests/io/formats/test_format.py::TestDataFrameFormatting::test_max_rows_fitted[50-30-10-10] - assert 30 == 10
FAILED pandas/tests/io/formats/test_format.py::TestDataFrameFormatting::test_max_rows_fitted[100-60-10-10] - assert 60 == 10
FAILED pandas/tests/io/formats/test_format.py::TestDataFrameFormatting::test_max_rows_fitted[61-60-10-10] - assert 60 == 10
=========================================================== 3 failed, 206 passed in 4.35s ============================================================
```

File: pandas/io/formats/format.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/io/formats/format.py:643]:
```
    def _is_screen_short(self, max_height) -> bool:
        return bool(self.max_rows == 0 and len(self.frame) > max_height)
```

If in terminal (max_rows==0) and if screen height (terminal height) is not enough to accommodate the dataframe entirely, then the screen is short.

Comment (ivanovmg) [reply pandas/io/formats/format.py:643]: If the logic can be improved, then please suggest. But my refactoring mainly consisted of removing comments and making methods with the same/similar names. Apparently got lost in some logic, but that was not covered by the tests properly.
I think that we can make a system test with the string representation as well, once we see that the present behavior is correct.

Comment (ivanovmg) [reply pandas/io/formats/format.py:643]: Code snippet before refactoring (regarding _is_screen_short).

```
            # Format only rows and columns that could potentially fit the
            # screen
            if max_cols == 0 and len(self.frame.columns) > w:
                max_cols = w
            if max_rows == 0 and len(self.frame) > h:
                max_rows = h
```

File: pandas/io/formats/format.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/io/formats/format.py:652]: I thought that some explanation should be added why there are still some changes to max_rows and comparison with frame length. Could be a comment, but a separate function with the docstring is better, isn't it.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls rebase

File: pandas/io/formats/format.py [hunk 4]
```diff
@@ -633,20 +633,30 @@ def _calc_max_cols_fitted(self) -> Optional[int]:
...
-        max_rows: Optional[int]
-        if self._is_screen_short(height):
-            max_rows = height
+        if self._is_in_terminal():
+            _, height = get_terminal_size()
+            if self.max_rows == 0:
+                # rows available to fill with actual data
+                return height - self._get_number_of_auxillary_rows()
+
+            max_rows: Optional[int]
```
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/io/formats/format.py:642]: I also think that splitting even further would be better.
```
if self._is_in_terminal():
    return self._get_max_rows_fitted_in_terminal()
else:
    return self._adjust_max_rows(self.max_rows)
```
Probably other PR?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@ivanovmg can you merge master.

@jorisvandenbossche if any comments.

**Comment by simonjayhawkins** [comment]
@ivanovmg can you merge master one more time and this can be merged cc @jorisvandenbossche

**Comment by ivanovmg** [comment]
> 
> 
> @ivanovmg can you merge master one more time and this can be merged cc @jorisvandenbossche

Done, it's green.

**Comment by jreback** [comment]
thanks @ivanovmg

**Pull request merged**